[DOC] ML-Commons Agent: Add documentation for PlanExecuteAndReflect Agent (#9743)
opensearch-project/documentation-website: _ml-commons-plugin/agents-tools/agents/plan-execute-reflect.md

## Creating a plan-execute-reflect agent

The following example request creates a plan-execute-reflect agent with three tools:

```json
POST /_plugins/_ml/agents/_register
{
  "name": "My Plan Execute Reflect Agent",
  "type": "plan_execute_and_reflect",
  "description": "Agent for dynamic task planning and reasoning",
  "llm": {
    "model_id": "YOUR_LLM_MODEL_ID",
    "parameters": {
      "prompt": "${parameters.question}"
    }
  },
  "memory": {
    "type": "conversation_index"
  },
  "parameters": {
    "_llm_interface": "YOUR_LLM_INTERFACE"
  },
  "tools": [
    { "type": "ListIndexTool" },
    { "type": "SearchIndexTool" },
    { "type": "IndexMappingTool" }
  ],
  "app_type": "os_chat"
}
```

It is important to provide thorough descriptions of the tools so that the LLM can decide in which situations to use those tools.
{: .tip}

For more information about the Register Agent API request fields, see [Request body fields]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/agent-apis/register-agent/#request-body-fields).

> ## Supported LLMs

Reviewer: OK then. How about introducing supported LLMs by saying: "The plan-execute-reflect agent provides built-in function calling interfaces for the following LLMs:" And then after the list, we can say "You can configure other LLMs to work with the plan-execute-reflect agent by configuring the internal parameters that `llm_interface` sets." Can we have an example of how to do this and what's involved? This will not be clear to the user as-is.

> ## Supported LLMs

Reviewer: Understood. I would remove the mentions of configuring other LLMs until that documentation/example is available. It might cause user frustration.

> ## Supported LLMs

Reviewer: I removed this line from register agent API table as well.

The plan-execute-reflect agent supports the following LLMs:

- [Anthropic Claude 3.7 model hosted on Amazon Bedrock](https://aws.amazon.com/bedrock/claude/)
- OpenAI GPT-4o model
- DeepSeek R1 model hosted on Amazon Bedrock

For a step-by-step tutorial on using a plan-execute-reflect agent, see [Building a plan-execute-reflect agent]({{site.url}}{{site.baseurl}}/tutorials/gen-ai/agents/build-plan-execute-reflect-agent/).

> To configure a plan-execute-reflect agent with a particular model, you need to modify the connector in [Step 1(a): Create a connector]({{site.url}}{{site.baseurl}}/tutorials/gen-ai/agents/build-plan-execute-reflect-agent/#step-1a-create-a-connector) and providing a model-specific `llm_interface` parameter in [Step 2: Create an agent]({{site.url}}{{site.baseurl}}/tutorials/gen-ai/agents/build-plan-execute-reflect-agent/#step-2-create-an-agent): 

Reviewer: ```suggestion
To configure a plan-execute-reflect agent with a particular model, you need to modify the connector in [Step 1(a): Create a connector]({{site.url}}{{site.baseurl}}/tutorials/gen-ai/agents/build-plan-execute-reflect-agent/#step-1a-create-a-connector) and provide a model-specific `llm_interface` parameter in [Step 2: Create an agent]({{site.url}}{{site.baseurl}}/tutorials/gen-ai/agents/build-plan-execute-reflect-agent/#step-2-create-an-agent): 
```

```json
"parameters": {
  "_llm_interface": "bedrock/converse/claude"
},
```

For valid values of the `_llm_interface` field, see [Request body fields]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/agent-apis/register-agent/#request-body-fields).

The following examples provide the connector and agent creation requests for the supported models.

> ### Anthropic Claude

Reviewer: ```suggestion
### Anthropic Claude on Amazon Bedrock
```

      "type": "IndexMappingTool"
    }
  ],
}
```
{% include copy-curl.html %}

### OpenAI GPT-4o

To create a connector for an OpenAI GPT-4o model, use the following request: 

```json
POST /_plugins/_ml/connectors/_create
{
    "name": "My openai connector: gpt-4",
    "description": "The connector to openai chat model",
    "version": 1,
    "protocol": "http",
    "parameters": {
        "model": "gpt-4o"
    },
    "credential": {
        "openAI_key": "your_open_ai_key"
    },
    "actions": [
        {
        "action_type": "predict",
        "method": "POST",
        "url": "https://api.openai.com/v1/chat/completions",
        "headers": {
            "Authorization": "Bearer ${credential.openAI_key}"
        },
        "request_body": "{ \"model\": \"${parameters.model}\", \"messages\": [{\"role\":\"developer\",\"content\":\"${parameters.system_prompt}\"},${parameters._chat_history:-}{\"role\":\"user\",\"content\":\"${parameters.prompt}\"}${parameters._interactions:-}]${parameters.tool_configs:-} }"
        }
    ]
}
```
{% include copy-curl.html %}

Then register the model and register an agent, specifying `openai/v1/chat/completions` in the `_llm_interface` field.

> ### Deepseek R1

Reviewer: ```suggestion
### Deepseek R1 on Amazon Bedrock
```

```
{% include copy.html %}

## Default prompts

The plan-execute-reflect agent uses the following predefined prompts. You can customize the prompts by providing new ones in the following ways:

- During agent registration in the `parameters` object
- Dynamically during agent execution

### Planner template and prompt 

The following template is used to ask the LLM to devise a plan for the given task:

```json
${parameters.planner_prompt} \n Objective: ${parameters.user_prompt} \n ${parameters.plan_execute_reflect_response_format}
```

The following prompt is used to ask the LLM to devise a plan for the given task:

```
For the given objective, come up with a simple step by step plan. This plan should involve individual tasks, that if executed correctly will yield the correct answer. Do not add any superfluous steps. The result of the final step should be the final answer. Make sure that each step has all the information needed - do not skip steps. At all costs, do not execute the steps. You will be told when to execute the steps.
```

> ### Planner prompt with a history template

Reviewer: @pyek-bot In each section, we can specify the parameter. For your example, the first sentence in the `Planner prompt` section will be  "To create a custom planner prompt, modify the `planner_prompt` parameter."

---
layout: default
title: Plan-execute-reflect agents
has_children: false
has_toc: false
nav_order: 40
parent: Agents
grand_parent: Agents and tools
---

# Plan-execute-reflect agents
**Introduced 3.0**
{: .label .label-purple }

This is an experimental feature and is not recommended for use in a production environment. For updates on the progress of the feature or if you want to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/ml-commons/issues/3745).    
{: .warning}

> Plan-execute-reflect agents are designed for solving complex tasks that require iterative reasoning and step-by-step execution. These agents use one LLM (the _planner_) to create and update a plan and another LLM (or the same one by default) to execute each individual step using a built-in conversational agent.

Reviewer 2: ```suggestion
Plan-execute-reflect agents are designed to solve complex tasks that require iterative reasoning and step-by-step execution. These agents use one large language model (LLM)---the _planner_---to create and update a plan and another LLM (or the same one by default) to execute each individual step using a built-in conversational agent.
```

```json
POST /_plugins/_ml/agents/_register
{
  "name": "My Plan Execute Reflect Agent",
  "type": "plan_execute_and_reflect",
  "description": "Agent for dynamic task planning and reasoning",
  "llm": {
    "model_id": "YOUR_LLM_MODEL_ID",
    "parameters": {
      "prompt": "${parameters.question}"
    }
  },
  "memory": {
    "type": "conversation_index"
  },
  "parameters": {
    "_llm_interface": "YOUR_LLM_INTERFACE"
  },
  "tools": [
    { "type": "ListIndexTool" },
    { "type": "SearchIndexTool" },
    { "type": "IndexMappingTool" }
  ],
  "app_type": "os_chat"
}
```

It is important to provide thorough descriptions of the tools so that the LLM can decide in which situations to use those tools.
{: .tip}

For more information about the Register Agent API request fields, see [Request body fields]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/agent-apis/register-agent/#request-body-fields).

## Supported LLMs

The plan-execute-reflect agent provides built-in function calling interfaces for the following LLMs:

- [Anthropic Claude 3.7 model hosted on Amazon Bedrock](https://aws.amazon.com/bedrock/claude/)
- OpenAI GPT-4o model
- DeepSeek-R1 model hosted on Amazon Bedrock

> To request default support for an LLM, create a feature request issue in [ml-commons](https://github.com/opensearch-project/ml-commons/issues).

Reviewer: ```suggestion
To request default support for an LLM, [create a feature request issue in the ML Commons repository](https://github.com/opensearch-project/ml-commons/issues).
```

      "type": "IndexMappingTool"
    }
  ]
}
```
{% include copy-curl.html %}

### OpenAI GPT-4o

To create a connector for an OpenAI GPT-4o model, use the following request: 

```json
POST /_plugins/_ml/connectors/_create
{
    "name": "My openai connector: gpt-4",
    "description": "The connector to openai chat model",
    "version": 1,
    "protocol": "http",
    "parameters": {
        "model": "gpt-4o"
    },
    "credential": {
        "openAI_key": "your_open_ai_key"
    },
    "actions": [
        {
        "action_type": "predict",
        "method": "POST",
        "url": "https://api.openai.com/v1/chat/completions",
        "headers": {
            "Authorization": "Bearer ${credential.openAI_key}"
        },
        "request_body": "{ \"model\": \"${parameters.model}\", \"messages\": [{\"role\":\"developer\",\"content\":\"${parameters.system_prompt}\"},${parameters._chat_history:-}{\"role\":\"user\",\"content\":\"${parameters.prompt}\"}${parameters._interactions:-}]${parameters.tool_configs:-} }"
        }
    ]
}
```
{% include copy-curl.html %}

Then register the model and register an agent, specifying `openai/v1/chat/completions` in the `_llm_interface` field.

> ### Deepseek R1 on Amazon Bedrock

Reviewer 2: ```suggestion
### Deepseek-R1 on Amazon Bedrock
```
